### ISSUE TITLE
[MThreads] Native upsample backward reference uses NEAREST and masks an incorrect FlagGems implementation

### ISSUE THREAD

**Issue by winfan1314** [FlagOpen/FlagGems#6207]
## Describe the bug

On the tested MThreads stack, the native MUSA implementations of several
`aten::upsample_*_backward` operators do not execute the interpolation mode
requested by the ATen entry point. The shared TorchMUSA backward wrapper passes
legacy `NEAREST` to muDNN in its 1D, 2D, and 3D branches, even when the caller
instantiated it with `LINEAR`, `BICUBIC`, or `NEAREST_EXACT`.

This has two consequences for FlagGems development:

1. The MUSA native operator is not a valid device accuracy or performance
   reference for linear/bilinear/bicubic/trilinear/nearest-exact backward on
   this release.
2. The current MThreads specialization of `upsample_linear1d_backward` in
   FlagGems master, introduced by #6175, reproduces the incorrect native
   nearest-style behavior. It can pass tests that use the broken device native
   operator as the reference while failing the requested PyTorch linear
   semantics against a CPU/opmath reference.

This report is limited to the exact software version below. A later TorchMUSA
binary may behave differently and should be revalidated independently.

## Environment

- Device: MTT S5000
- Python: 3.10.20
- PyTorch: 2.9.1, build commit
  `d38164a545b4a4e4e0cf73ce67173f70574890b6`
- TorchMUSA distribution: 2.9.1
- TorchMUSA module version: `2.9.1+a18d871`
- TorchMUSA reported commit:
  `a18d871c7fd549bcd6b7b927e888d7666bdc666b`
- Tested FlagGems master:
  `fcdd6ee973bf277b7852ce465d044c85674b375b`
- Current MThreads specialization from #6175:
  `88000969a3`

The installed TorchMUSA `UpSample.cpp` and `Interpolate.cpp` were byte-for-byte
identical to the public TorchMUSA v2.9.1 files at
`467bb873388939c646ac4525f463bcdcc3eb570e`.

## Minimal native-only reproduction

This process imports only `torch` and `torch_musa`. It also asserts that
FlagGems has not been imported.

```python
import sys

import torch
import torch_musa

assert not any(
    name == "flag_gems" or name.startswith("flag_gems.")
    for name in sys.modules
)

grad = torch.tensor(
    [[[-1.25, 0.5, 2.0, -0.75, 1.5, -2.25, 0.875]]],
    dtype=torch.float32,
)
output_size = [7]
input_size = [1, 1, 4]

cpu_linear = torch.ops.aten.upsample_linear1d_backward.default(
    grad.double(), output_size, input_size, False, None
).float()
cpu_nearest = torch.ops.aten.upsample_nearest1d_backward.default(
    grad.double(), output_size, input_size, None
).float()
actual = torch.ops.aten.upsample_linear1d_backward.default(
    grad.to("musa"), output_size, input_size, False, None
).cpu()

print("matches_linear", torch.equal(actual, cpu_linear))
print("matches_nearest", torch.equal(actual, cpu_nearest))
```

Observed result:

```text
matches_linear False
matches_nearest True
```

`align_corners=True` and the `grad_input` overload show the same mode mismatch.

## Native-only evidence

The expanded linear1d matrix used 54 fresh processes and 324 native MUSA calls.
Each process checked that FlagGems was absent from `sys.modules`. The matrix
covered:

- FP32, FP16, and BF16;
- identity, integer 2x upsampling, non-integer upsampling, downsampling,
  output width 1, and explicit-scale/size disagreement;
- ramp, impulse, and fixed-seed inputs;
- `align_corners=False/True`;
- allocating `default` and caller-provided `grad_input` overloads.

Results:

- 216/216 native linear1d outputs were bitwise equal to native legacy-nearest.
- 156 cases mathematically distinguished linear from nearest; 156/156 matched
  nearest bitwise and 0/156 matched linear.
- 108/108 `align_corners=False/True` pairs had identical output hashes.
- 36/36 explicit-scale linear calls matched the nearest mapping.
- 162/162 allocating/out pairs agreed, so this is not an overload problem.
- 18/18 profiler calls launched a MUSA device kernel. Within each dtype,
  nearest and every linear variant used the same
  `Interpolate1d ... Mode)0` kernel.

This rules out FlagGems interception, CPU fallback, schema/PrivateUse1
registration, host-to-device corruption, overload differences, and ordinary
low-precision accumulation noise.

The same installed release was independently checked for 2D backward:

- `upsample_bilinear2d_backward`: 24 numerical comparisons and 30 device
  kernel events; native bilinear matched nearest, and all device calls selected
  the nearest interpolation kernel template.
- `upsample_bicubic2d_backward`: 120 original-dtype samples; all matched nearest
  bitwise, while 96 mathematically distinguishable samples differed from cubic.
  Another 15 device calls showed nearest and both bicubic overloads/alignment
  modes using the same kernel name within each dtype.

Adjacent probes found:

- `_upsample_nearest_exact1d_backward` matched legacy nearest rather than exact
  nearest.
- `upsample_trilinear3d_backward` matched nearest3d for both alignment modes;
  the maximum absolute difference from CPU trilinear was 1.859375.
- Six FP64 nearest/linear default/out calls failed with
  `UpSampleNdBwdOut MUDNN failed in: RunBackward`; this is a separate dtype
  support boundary rather than evidence for the wrong-mode classification.

Nearest-exact 2D/3D and anti-aliased backward statically share the affected
wrapper, but they were not all exercised on hardware in this investigation.

## Source-level root cause in TorchMUSA

The ATen entry points instantiate the shared template correctly:

```cpp
UpSampleNdBwdOut<1, INTERPOLATE_MODE::LINEAR>(...);
UpSampleNdBwdOut<2, INTERPOLATE_MODE::LINEAR>(...);
UpSampleNdBwdOut<2, INTERPOLATE_MODE::BICUBIC>(...);
```

However, `UpSampleNdBwdOut<Nd, mode>` ignores `mode` in all three dimension
branches and hard-codes:

```cpp
::at::musa::SetInterpolate(
    op,
    ::musa::dnn::Interpolate::Mode::NEAREST,
    antialias,
    should_set_align_corners ? align_corners : false,
    scales);
```

Source references:

- TorchMUSA shared backward wrapper:
  https://github.com/MooreThreads/torch_musa/blob/467bb873388939c646ac4525f463bcdcc3eb570e/torch_musa/csrc/aten/ops/UpSample.cpp
- TorchMUSA descriptor setup, which faithfully forwards the mode it receives:
  https://github.com/MooreThreads/torch_musa/blob/467bb873388939c646ac4525f463bcdcc3eb570e/torch_musa/csrc/aten/mudnn/Interpolate.cpp
- PyTorch interpolation helpers:
  https://github.com/pytorch/pytorch/blob/v2.9.1/aten/src/ATen/native/UpSample.h

The proven defect therefore occurs before muDNN execution. This investigation
did not rebuild TorchMUSA, so correct behavior after passing a real
LINEAR/BICUBIC descriptor to the installed muDNN remains to be verified.

## Impact on the current FlagGems master implementation

#6175 replaced the previous MThreads implementation and explicitly modeled the
observed target native behavior. Its implementation:

- maps each `grad_output[dst]` with weight 1 to
  `floor((W_in / W_out) * dst)` in the generic paths;
- uses `grad[2*i] + grad[2*i+1]` for the exact 2x path;
- writes each downsampled gradient to one even input position and writes zero
  to the paired odd position for the 0.5x path;
- does not use `align_corners` to choose interpolation coordinates or weights.

Those operations are nearest-style accumulation, not linear-backward
accumulation into two neighboring input positions with interpolation weights.

The statement in #6175 that all 1080 accuracy cases passed used the MUSA device
native operator as the reference. Because both sides implemented the same wrong
mode, that result did not establish PyTorch linear semantics.

Using the CPU-reference test from #5066 against identical isolated source trees
on the same MTT S5000 produced:

| Implementation | Main matrix | BF16 cancellation regression | Total |
|---|---:|---:|---:|
| #5066 (`5071b3fe59`) | 1080/1080 passed | 1/1 passed | 1081/1081 passed |
| master with #6175 | 548/1080 passed | failed | 548 passed, 533 failed |

Of the 540 non-identity main cases, master failed 532:

- all 270 `align_corners=False` cases failed;
- all 270 cases in the non-contiguous-layout test group failed;
- scale 0.5/1.5/2.0/2.5/4.0 failed
  102/108, 108/108, 108/108, 107/108, and 107/108 cases;
- FP16/FP32/BF16 failed 178/180, 178/180, and 176/180 cases.

The other 540 passing cases were repeated `1 -> 1` identity edge cases. Only
eight non-identity cases passed, all small accidental coincidences. The separate
BF16 cancellation case failed with a maximum absolute difference of 0.796875
at an allowed tolerance of 0.02.

In a 168-case comprehensive kernel-latency A/B, the master implementation was
faster in 150 cases and had a 1.291x geometric-mean advantage over #5066. That
speed is not a valid optimization result: it comes from executing the wrong,
cheaper interpolation algorithm. Likewise, the native MUSA latency in this
release measures nearest behavior and must not be published as a linear,
bilinear, bicubic, or trilinear speedup baseline.

## Related FlagGems PRs

Direct current MThreads implementation and validation lineage:

- #5066: open SiliconFlow fix/optimization; provides the MThreads implementation
  that passes 1081/1081 CPU-reference cases and changes MThreads validation away
  from the invalid device native reference.
- #5779: previous merged MThreads specialization using linear interpolation
  windows; later replaced by #6175.
- #6175: current merged MThreads specialization that reproduces nearest-style
  native semantics and is the accuracy-regressing implementation described
  above.
- #5219: open generic correctness fix for dropped `align_corners=True` gradient
  terms at large ratios; relevant to the expected linear-backward semantics.

SiliconFlow operator/test history:

- #2069: earlier closed SiliconFlow submission for `upsample_linear1d_backward`.
- #2088: merged SiliconFlow PR that introduced the generic operator, tests, and
  benchmark.
- #3943: closed precursor of the Ascend specialization.
- #3944: merged SiliconFlow Ascend implementation and scale-handling/test update.
- #4864: closed aggregate SiliconFlow validation PR that contained the MThreads
  implementation, Ascend repair, and the MThreads reference-policy correction;
  its focused successor is #5066.

## Expected behavior and suggested fixes

1. TorchMUSA should pass the requested template `mode` into `SetInterpolate` in
   the 1D, 2D, and 3D backward branches. The alignment condition should match
   forward and apply to `LINEAR || BICUBIC`.
2. Until a rebuilt TorchMUSA package is verified, MThreads FlagGems accuracy
   tests for the affected backward operators should use a CPU/opmath reference,
   not the MUSA device native implementation.
3. The #6175 implementation in master should be replaced or corrected to
   implement true linear weights and both `align_corners` modes. #5066 provides
   a currently tested implementation and regression coverage.
4. Performance tests should report FlagGems kernel latency independently and
   mark the native latency/speedup as invalid for affected modes on this exact
   TorchMUSA release.
5. After rebuilding TorchMUSA, rerun both overloads across nearest,
   nearest-exact, linear, bilinear, bicubic, trilinear, and anti-aliased
   backward; both alignment modes; explicit and derived scales; FP32/FP16/BF16;
   and identity/non-identity shapes before restoring device-native reference
   comparisons.

## CI reproduction with a comment-only trigger PR

PR #6208 provides an independently inspectable CI reproduction without changing runtime or test semantics. It adds one comment to `tests/test_upsample_linear1d.py`, which places that file in `CHANGED_FILES`, and one comment to the MThreads backend implementation, which makes triage apply `vendor/MooreThreads` and schedules `backend-tests (mthreads-musa520)`.

The backend workflow runs the same changed test file twice via `tools/test-op.sh`:

1. The normal device-reference phase compares FlagGems with native MUSA. It reported **1107 passed, 120 skipped**, so the current wrong nearest-style implementation is masked by the equally wrong native reference.
2. The quick CPU-reference phase reruns the file with `--ref=cpu --quick`. It failed immediately on `test_upsample_linear1d_backward[False-contiguous-False-0.5-dtype0-shape0]`: FlagGems returned `[-1.3203, 0.0000]`, while the CPU linear reference returned `[-0.6602, -0.6602]`; 2/2 elements mismatched, with maximum absolute difference 0.66015625 at tolerance 0.02.

This A/B occurs in one job, on one checkout and runner, with only the reference path changing. It directly demonstrates both sides of the issue: device-native validation passes the incorrect implementation, while CPU/opmath validation exposes the linear-semantics error.

Reproduction links:

- Comment-only trigger PR: https://github.com/flagos-ai/FlagGems/pull/6208
- MThreads failing job: https://github.com/flagos-ai/FlagGems/actions/runs/34635632532/job/103382902009
- Trigger commit: https://github.com/flagos-ai/FlagGems/commit/054d44954366325272096b39038ccc21b0dcc8a0
